File: scripts/generateCitizenImages.ts
const axios = require('axios');
const fs = require('fs');
const path = require('path');
const dotenv = require('dotenv');

dotenv.config();

// Define citizen interface
interface Citizen {
  id: string;
  socialClass: 'Patrician' | 'Cittadini' | 'Popolani' | 'Laborer';
  firstName: string;
  lastName: string;
  description: string;
  imagePrompt: string;
  imageUrl?: string;
  wealth: number;
  createdAt: string;
}

// Ideogram API configuration
const IDEOGRAM_API_KEY = process.env.IDEOGRAM_API_KEY;

// Path to citizens data and image directory
const CITIZENS_DATA_PATH = path.join(process.cwd(), 'data', 'citizens.json');
const CITIZENS_IMAGE_DIR = path.join(process.cwd(), 'public', 'images', 'citizens');

// Ensure the images directory exists
if (!fs.existsSync(CITIZENS_IMAGE_DIR)) {
  fs.mkdirSync(CITIZENS_IMAGE_DIR, { recursive: true });
}

// Load citizens data
function loadCitizens(): Citizen[] {
  try {
    const data = fs.readFileSync(CITIZENS_DATA_PATH, 'utf8');
    return JSON.parse(data);
  } catch (error) {
    console.error('Error loading citizens data:', error);
    return [];
  }
}

// Save updated citizens data
function saveCitizens(citizens: Citizen[]): void {
  try {
    fs.writeFileSync(CITIZENS_DATA_PATH, JSON.stringify(citizens, null, 2));
    console.log(`Saved updated citizens data to ${CITIZENS_DATA_PATH}`);
  } catch (error) {
    console.error('Error saving citizens data:', error);
  }
}

// Enhance image prompt with style guidelines
function enhanceImagePrompt(citizen: Citizen): string {
  const basePrompt = citizen.imagePrompt;
  
  // Add style guidelines based on social class
  let styleAddition = '';
  
  switch (citizen.socialClass) {
    case 'Patrician':
      styleAddition = 'Renaissance portrait style with realistic details. 3/4 view portrait composition with Rembrandt lighting. Rich color palette with deep reds and gold tones. --ar 1:1';
      break;
    case 'Cittadini':
      styleAddition = 'Renaissance portrait style with realistic details. 3/4 view portrait composition with warm Rembrandt lighting. Warm amber tones. --ar 1:1';
      break;
    case 'Popolani':
      styleAddition = 'Renaissance portrait style with realistic details. 3/4 view portrait composition with directional lighting. Muted earth tones. --ar 1:1';
      break;
    case 'Laborer':
      styleAddition = 'Renaissance portrait style with realistic details. 3/4 view portrait composition with natural lighting. Subdued color palette. --ar 1:1';
      break;
    default:
      styleAddition = 'Renaissance portrait style with realistic details. 3/4 view portrait composition. --ar 1:1';
  }
  
  // Combine original prompt with style guidelines
  return `${basePrompt} ${styleAddition}`;
}

// Generate image using Ideogram API
async function generateImage(prompt: string, citizenId: string): Promise<string | null> {
  try {
    console.log(`Sending prompt to Ideogram API: ${prompt.substring(0, 100)}...`);
    
    // Create form data for multipart request
    const FormData = require('form-data');
    const form = new FormData();
    
    // Add required parameters
    form.append('prompt', prompt);
    form.append('style_type', 'REALISTIC');
    form.append('rendering_speed', 'DEFAULT');
    
    const response = await axios.post(
      'https://api.ideogram.ai/v1/ideogram-v3/generate',
      form,
      {
        headers: {
          ...form.getHeaders(),
          'Api-Key': IDEOGRAM_API_KEY
        }
      }
    );
    
    // Extract image URL from response
    const imageUrl = response.data.data[0].url;
    
    // Download the image
    const imageResponse = await axios.get(imageUrl, { responseType: 'arraybuffer' });
    const imagePath = path.join(CITIZENS_IMAGE_DIR, `${citizenId}.jpg`);
    fs.writeFileSync(imagePath, imageResponse.data);
    
    console.log(`Generated and saved image for citizen ${citizenId}`);
    
    return `/images/citizens/${citizenId}.jpg`;
  } catch (error) {
    console.error(`Error generating image for citizen ${citizenId}:`, error);
    return null;
  }
}

// Main function to generate images for all citizens
async function generateCitizenImages(limit: number = 0): Promise<void> {
  const citizens = loadCitizens();
  let updatedCount = 0;
  let processedCount = 0;
  
  for (let i = 0; i < citizens.length; i++) {
    const citizen = citizens[i];
    
    // Skip if citizen already has an image
    if (citizen.imageUrl) {
      console.log(`Citizen ${citizen.id} already has an image, skipping...`);
      continue;
    }
    
    // Stop if we've reached the limit (if specified)
    if (limit > 0 && processedCount >= limit) {
      console.log(`Reached limit of ${limit} images, stopping.`);
      break;
    }
    
    console.log(`Generating image for citizen ${i+1}/${citizens.length}: ${citizen.firstName} ${citizen.lastName}`);
    
    // Enhance the image prompt with style guidelines
    const enhancedPrompt = enhanceImagePrompt(citizen);
    
    // Generate the image
    const imageUrl = await generateImage(enhancedPrompt, citizen.id);
    
    if (imageUrl) {
      // Update citizen with image URL
      citizens[i].imageUrl = imageUrl;
      updatedCount++;
      
      // Save after each successful generation to avoid losing progress
      saveCitizens(citizens);
    }
    
    processedCount++;
    
    // Add a delay to avoid rate limiting
    await new Promise(resolve => setTimeout(resolve, 3000));
  }
  
  console.log(`Generated images for ${updatedCount} citizens out of ${processedCount} processed`);
}

// Command line interface
async function main(): Promise<void> {
  const args = process.argv.slice(2);
  const limit = args.length > 0 ? parseInt(args[0]) : 0;
  
  if (args.length > 0 && isNaN(limit)) {
    console.error('Please provide a valid limit (positive integer)');
    process.exit(1);
  }
  
  console.log(`Starting citizen image generation${limit > 0 ? ` with limit of ${limit} images` : ''}`);
  await generateCitizenImages(limit);
}

// Run the script
main().catch(console.error);

File: backend/engine/generate_citizen_images.py
#!/usr/bin/env python3
"""
Generate images for citizens using the Ideogram API.

This script:
1. Fetches citizens from Airtable that need images
2. Generates images using the Ideogram API
3. Saves the images to the public/images/citizens directory
4. Updates the citizen records in Airtable with the image URLs

It can be run directly or imported and used by other scripts.
"""

import os
import sys
import logging
import argparse
import json
import time
import requests
from typing import Dict, List, Optional, Any
from pyairtable import Api, Table
from dotenv import load_dotenv

# Set up logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
log = logging.getLogger("generate_citizen_images")

# Load environment variables
load_dotenv()

# Constants
CITIZENS_IMAGE_DIR = os.path.join(os.getcwd(), 'public', 'images', 'citizens')

# Ensure the images directory exists
os.makedirs(CITIZENS_IMAGE_DIR, exist_ok=True)

def initialize_airtable():
    """Initialize Airtable connection."""
    api_key = os.environ.get('AIRTABLE_API_KEY')
    base_id = os.environ.get('AIRTABLE_BASE_ID')
    
    if not api_key or not base_id:
        log.error("Missing Airtable credentials. Set AIRTABLE_API_KEY and AIRTABLE_BASE_ID environment variables.")
        sys.exit(1)
    
    try:
        # Return a dictionary of table objects using pyairtable
        return {
            'citizens': Table(api_key, base_id, 'CITIZENS')
        }
    except Exception as e:
        log.error(f"Failed to initialize Airtable: {e}")
        sys.exit(1)

def fetch_citizens_needing_images(tables) -> List[Dict]:
    """Fetch citizens from Airtable that need images."""
    log.info("Fetching citizens from Airtable that need images...")
    
    try:
        # Get citizens without an ImageUrl field or with empty ImageUrl field
        formula = "OR({ImageUrl} = '', {ImageUrl} = BLANK())"
        citizens = tables['citizens'].all(formula=formula)
        
        log.info(f"Found {len(citizens)} citizens needing images")
        return citizens
    except Exception as e:
        log.error(f"Error fetching citizens needing images: {e}")
        return []

def enhance_image_prompt(citizen: Dict) -> str:
    """Enhance image prompt with style guidelines based on social class."""
    base_prompt = citizen['fields'].get('ImagePrompt', '')
    
    # Add style guidelines based on social class
    social_class = citizen['fields'].get('SocialClass', '')
    
    style_addition = ''
    
    if social_class == 'Nobili':
        style_addition = 'Renaissance portrait style with realistic details. 3/4 view portrait composition with Rembrandt lighting. Rich color palette with deep reds and gold tones. Ornate clothing with fine details. Aristocratic bearing and confident expression. Venetian palazzo background with marble columns.'
    elif social_class == 'Cittadini':
        style_addition = 'Renaissance portrait style with realistic details. 3/4 view portrait composition with warm Rembrandt lighting. Warm amber tones. Quality clothing with some decorative elements. Intelligent and dignified expression. Venetian merchant office or study background.'
    elif social_class == 'Popolani':
        style_addition = 'Renaissance portrait style with realistic details. 3/4 view portrait composition with directional lighting. Muted earth tones. Practical, well-made clothing. Hardworking and capable expression. Workshop or marketplace background with tools of their trade.'
    elif social_class in ['Facchini', 'Laborer']:
        style_addition = 'Renaissance portrait style with realistic details. 3/4 view portrait composition with natural lighting. Subdued color palette. Simple, functional clothing. Weather-worn features with determined expression. Venetian docks or working environment background.'
    else:
        style_addition = 'Renaissance portrait style with realistic details. 3/4 view portrait composition with balanced lighting. Clothing and setting appropriate to their profession in Renaissance Venice.'
    
    # Add citizen's name and profession if available
    first_name = citizen['fields'].get('FirstName', '')
    last_name = citizen['fields'].get('LastName', '')
    
    # Combine original prompt with style guidelines
    enhanced_prompt = f"{base_prompt} {style_addition}"
    
    # Add photorealistic quality directive
    enhanced_prompt += " Photorealistic quality, highly detailed facial features, historically accurate."
    
    return enhanced_prompt

def update_airtable_image_url(tables, citizen_id: str, image_url: str) -> bool:
    """Update Airtable with image URL."""
    log.info(f"Updating Airtable record for citizen {citizen_id} with image URL: {image_url}")
    
    try:
        # Update the record with the new image URL
        tables['citizens'].update(citizen_id, {
            "ImageUrl": image_url
        })
        
        log.info(f"Successfully updated Airtable record for citizen {citizen_id}")
        return True
    except Exception as e:
        log.error(f"Error updating Airtable record for citizen {citizen_id}: {e}")
        return False

def generate_image(prompt: str, citizen_id: str) -> Optional[str]:
    """Generate image using Ideogram API."""
    log.info(f"Sending prompt to Ideogram API: {prompt[:100]}...")
    
    # Get Ideogram API key from environment
    ideogram_api_key = os.environ.get('IDEOGRAM_API_KEY')
    if not ideogram_api_key:
        log.error("IDEOGRAM_API_KEY environment variable is not set")
        return None
    
    try:
        # Call the Ideogram API
        response = requests.post(
            "https://api.ideogram.ai/v1/ideogram-v3/generate",
            headers={
                "Api-Key": ideogram_api_key,
                "Content-Type": "application/json"
            },
            json={
                "prompt": prompt,
                "style_type": "REALISTIC",
                "rendering_speed": "DEFAULT",
                "model":"V_3"
            }
        )
        
        if response.status_code != 200:
            log.error(f"Error from Ideogram API: {response.status_code} {response.text}")
            return None
        
        # Extract image URL from response
        result = response.json()
        image_url = result.get("data", [{}])[0].get("url", "")
        
        if not image_url:
            log.error("No image URL in response")
            return None
        
        # Download the image
        image_response = requests.get(image_url, stream=True)
        if not image_response.ok:
            log.error(f"Failed to download image: {image_response.status_code} {image_response.reason}")
            return None
        
        # Save the image to the public folder
        image_path = os.path.join(CITIZENS_IMAGE_DIR, f"{citizen_id}.jpg")
        with open(image_path, 'wb') as f:
            for chunk in image_response.iter_content(chunk_size=8192):
                f.write(chunk)
        
        log.info(f"Generated and saved image for citizen {citizen_id}")
        
        # Create the public URL path
        public_image_url = f"/images/citizens/{citizen_id}.jpg"
        
        return public_image_url
    except Exception as e:
        log.error(f"Error generating image for citizen {citizen_id}: {e}")
        return None

def process_citizen(tables, citizen: Dict) -> bool:
    """Process a single citizen to generate an image."""
    citizen_id = citizen['id']
    citizen_name = f"{citizen['fields'].get('FirstName', '')} {citizen['fields'].get('LastName', '')}"
    
    log.info(f"Processing citizen: {citizen_name} (ID: {citizen_id})")
    
    # Get the image prompt
    image_prompt = citizen['fields'].get('ImagePrompt', '')
    if not image_prompt:
        log.warning(f"No image prompt for citizen {citizen_id}")
        return False
    
    # Enhance the prompt
    enhanced_prompt = enhance_image_prompt(citizen)
    
    # Generate the image
    image_url = generate_image(enhanced_prompt, citizen_id)
    if not image_url:
        log.warning(f"Failed to generate image for citizen {citizen_id}")
        return False
    
    # Update Airtable with the image URL
    success = update_airtable_image_url(tables, citizen_id, image_url)
    
    return success

def process_specific_citizen(tables, citizen_id: str, image_prompt: str) -> bool:
    """Process a specific citizen with the given ID and prompt."""
    log.info(f"Processing specific citizen: {citizen_id}")
    
    if not image_prompt:
        log.warning(f"No image prompt provided for citizen {citizen_id}")
        return False
    
    # Generate the image
    image_url = generate_image(image_prompt, citizen_id)
    if not image_url:
        log.warning(f"Failed to generate image for citizen {citizen_id}")
        return False
    
    # Update Airtable with the image URL
    success = update_airtable_image_url(tables, citizen_id, image_url)
    
    return success

def generate_citizen_images(limit: int = 0):
    """Generate images for citizens that need them."""
    tables = initialize_airtable()
    
    # Check if we're processing a specific citizen from command line
    if len(sys.argv) > 2 and sys.argv[1] == '--citizen-id':
        citizen_id = sys.argv[2]
        
        # Check if we have a temp file with citizen data
        if os.path.exists('temp_citizen_image.json'):
            try:
                with open('temp_citizen_image.json', 'r') as f:
                    citizen_data = json.load(f)
                
                success = process_specific_citizen(tables, citizen_id, citizen_data.get('imagePrompt', ''))
                log.info(f"Processed citizen {citizen_id} with result: {'success' if success else 'failed'}")
                return
            except Exception as e:
                log.error(f"Error processing citizen from temp file: {e}")
        
        log.error(f"No temp file found for citizen {citizen_id}")
        return
    
    # Normal flow - fetch citizens from Airtable
    citizens = fetch_citizens_needing_images(tables)
    
    if not citizens:
        log.info("No citizens found that need images. Exiting.")
        return
    
    log.info(f"Found {len(citizens)} citizens that need images")
    
    updated_count = 0
    processed_count = 0
    
    for i, citizen in enumerate(citizens):
        # Stop if we've reached the limit (if specified)
        if limit > 0 and processed_count >= limit:
            log.info(f"Reached limit of {limit} images, stopping.")
            break
        
        citizen_name = f"{citizen['fields'].get('FirstName', '')} {citizen['fields'].get('LastName', '')}"
        log.info(f"Generating image for citizen {i+1}/{len(citizens)}: {citizen_name}")
        
        success = process_citizen(tables, citizen)
        
        if success:
            updated_count += 1
        
        processed_count += 1
        
        # Add a delay to avoid rate limiting
        time.sleep(3)
    
    log.info(f"Generated images for {updated_count} citizens out of {processed_count} processed")

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Generate images for citizens")
    parser.add_argument("limit", nargs="?", type=int, default=0, help="Maximum number of images to generate (0 for unlimited)")
    parser.add_argument("--citizen-id", help="Generate image for a specific citizen ID")
    
    args = parser.parse_args()
    
    generate_citizen_images(args.limit)
